audio-worklet/design-pattern/lib/wa2.js

```diff
@@ -0,0 +1,289 @@
+   *
+   * @return {number} Buffer length in frames.
+   */
+  get maxChannelCount() {
+    return this._isInitialized ? this._maxChannelCount : null;
+  }
+
+  /**
+   * Returns a Float32Array object for a given channel index. If the channel
+   * index is undefined, it returns the reference to the base array of channel
+   * data.
+   *
+   * @param  {number|undefined} channelIndex Channel index.
+   * @return {Float32Array|Array[Float32Array]} a channel data array or an
+   * array of channel data.
+   */
+  getChannelData(channelIndex) {
+    console.assert(this._isInitialized);
+    console.assert(typeof channelIndex === 'undefined' ||
+                   channelIndex < this._channelCount);
+
+    return typeof channelIndex === 'undefined'
+        ? this._channelData : this._channelData[channelIndex];
+  }
+
+  /**
+   * Returns the base address of the allocated memory space in the WASM heap.
+   *
+   * @return {number} WASM Heap address.
+   */
+  getHeap() {
+    console.assert(this._isInitialized);
+
+    return this._dataPtr;
+  }
+
+  /**
+   * Frees the allocated memory space in the WASM heap.
+   *
+   * @return {[type]} [description]
+   */
+  free() {
+    console.assert(this._isInitialized);
+
+    this._isInitialized = false;
+    Module._free(this._dataPtr);
+    Module._free(this._pointerArrayPtr);
+    this._channelData = null;
+  }
+
+}  // class HeapAudioBuffer
+
+
+/**
+ * A JS FIFO implementation for the AudioWorklet. 3 assumptions for the
+ * simpler operation:
+ *  1. the push and the pull operation are done by 128 frames. (Web Audio
+ *    API's render quantum size in the speficiation)
+ *  2. the channel count of input/output cannot be changed dynamically.
+ *    The AudioWorkletNode should be configured with the `.channelCount = k`
+ *    (where k is the channel count you want) and
+ *    `.channelCountMode = explicit`.
+ *  3. This is for the single-thread operation. (obviously)
+ *
+ * @class
+ */
+class RingBuffer {
+
+  /**
+   * @constructor
+   * @param  {number} channelCount Buffer channel count.
+   * @param  {number} length Buffer length in frames.
+   * @return {RingBuffer}
+   */
+  constructor(length, channelCount) {
+    this._ri = 0;
+    this._wi = 0;
+    this._framesAvailable = 0;
+
+    this._channelCount = channelCount;
+    this._length = length;
+    this._channelData = [];
+    for (let i = 0; i < this._channelCount; ++i) {
+      this._channelData[i] = new Float32Array(length);
+    }
+  }
+
+  /**
+   * Getter for Available frames in buffer.
+   *
+   * @return {number} Available frames in buffer.
+   */
+  get framesAvailable() {
+    return this._framesAvailable;
+  }
+
+  /**
+   * Push a sequence of Float32Arrays to buffer.
+   *
+   * @param  {array} arraySequence A sequence of Float32Arrays.
+   */
+  push(arraySequence) {
+    console.assert(arraySequence.length === this._channelCount);
+    console.assert(arraySequence[0].length <= this._length);
+
```

Comment:
Need to check that the Float32Arrays all have the same length?

Reply:
This is targeted for the AudioWorklet, so no.

Reply:
As we discussed, I am removing all "assert"s from the code. Some comments would serve the purpose of the example code better.